File: go/src/step5_tco/step5_tco.go
package main

import (
    "fmt"
    "strings"
    "errors"
)

import (
    "readline"
    . "types"
    "reader"
    "printer"
    . "env"
    "core"
)

// read
func READ(str string) (MalType, error) {
    return reader.Read_str(str)
}

// eval
func eval_ast(ast MalType, env EnvType) (MalType, error) {
    //fmt.Printf("eval_ast: %#v\n", ast)
    if Symbol_Q(ast) {
        return env.Get(ast.(Symbol))
    } else if List_Q(ast) {
        lst := []MalType{}
        for _, a := range ast.(List).Val {
            exp, e := EVAL(a, env)
            if e != nil { return nil, e }
            lst = append(lst, exp)
        }
        return List{lst,nil}, nil
    } else if Vector_Q(ast) {
        lst := []MalType{}
        for _, a := range ast.(Vector).Val {
            exp, e := EVAL(a, env)
            if e != nil { return nil, e }
            lst = append(lst, exp)
        }
        return Vector{lst,nil}, nil
    } else if HashMap_Q(ast) {
        m := ast.(HashMap)
        new_hm := HashMap{map[string]MalType{},nil}
        for k, v := range m.Val {
            ke, e1 := EVAL(k, env)
            if e1 != nil { return nil, e1 }
            if _, ok := ke.(string); !ok {
                return nil, errors.New("non string hash-map key")
            }
            kv, e2 := EVAL(v, env)
            if e2 != nil { return nil, e2 }
            new_hm.Val[ke.(string)] = kv
        }
        return new_hm, nil
    } else {
        return ast, nil
    }
}

func EVAL(ast MalType, env EnvType) (MalType, error) {
    for {

    //fmt.Printf("EVAL: %v\n", printer.Pr_str(ast, true))
    switch ast.(type) {
    case List: // continue
    default:   return eval_ast(ast, env)
    }

    // apply list
    a0 := ast.(List).Val[0]
    var a1 MalType = nil; var a2 MalType = nil
    switch len(ast.(List).Val) {
    case 1:
        a1 = nil; a2 = nil
    case 2:
        a1 = ast.(List).Val[1]; a2 = nil
    default:
        a1 = ast.(List).Val[1]; a2 = ast.(List).Val[2]
    }
    a0sym := "__<*fn*>__"
    if Symbol_Q(a0) { a0sym = a0.(Symbol).Val } 
    switch a0sym {
    case "def!":
        res, e := EVAL(a2, env)
        if e != nil { return nil, e }
        return env.Set(a1.(Symbol), res), nil
    case "let*":
        let_env, e := NewEnv(env, nil, nil)
        if e != nil { return nil, e }
        arr1, e := GetSlice(a1)
        if e != nil { return nil, e }
        for i := 0; i < len(arr1); i+=2 {
            if !Symbol_Q(arr1[i]) {
                return nil, errors.New("non-symbol bind value")
            }
            exp, e := EVAL(arr1[i+1], let_env)
            if e != nil { return nil, e }
            let_env.Set(arr1[i].(Symbol), exp)
        }
        ast = a2
        env = let_env
    case "do":
        lst := ast.(List).Val
        _, e := eval_ast(List{lst[1:len(lst)-1],nil}, env) 
        if e != nil { return nil, e }
        if len(lst) == 1 { return nil, nil }
        ast = lst[len(lst)-1]
    case "if":
        cond, e := EVAL(a1, env)
        if e != nil { return nil, e }
        if cond == nil || cond == false {
            if len(ast.(List).Val) >= 4 {
                ast = ast.(List).Val[3]
            } else {
                return nil, nil
            }
        } else {
            ast = a2
        }
    case "fn*":
        fn := MalFunc{EVAL, a2, env, a1, false, NewEnv, nil}
        return fn, nil
    default:
        el, e := eval_ast(ast, env)
        if e != nil { return nil, e }
        f := el.(List).Val[0]
        if MalFunc_Q(f) {
            fn := f.(MalFunc)
            ast = fn.Exp
            env, e = NewEnv(fn.Env, fn.Params, List{el.(List).Val[1:],nil})
            if e != nil { return nil, e }
        } else {
            fn, ok := f.(Func)
            if !ok { return nil, errors.New("attempt to call non-function") }
            return fn.Fn(el.(List).Val[1:])
        }
    }

    } // TCO loop
}

// print
func PRINT(exp MalType) (string, error) {
    return printer.Pr_str(exp, true), nil
}


var repl_env, _ = NewEnv(nil, nil, nil)

// repl
func rep(str string) (MalType, error) {
    var exp MalType
    var res string
    var e error
    if exp, e = READ(str); e != nil { return nil, e }
    if exp, e = EVAL(exp, repl_env); e != nil { return nil, e }
    if res, e = PRINT(exp); e != nil { return nil, e }
    return res, nil
}

func main() {
    // core.go: defined using go
    for k, v := range core.NS {
        repl_env.Set(Symbol{k}, Func{v.(func([]MalType)(MalType,error)),nil})
    }

    // core.mal: defined using the language itself
    rep("(def! not (fn* (a) (if a false true)))")

    // repl loop
    for {
        text, err := readline.Readline("user> ")
        text = strings.TrimRight(text, "\n");
        if (err != nil) {
            return
        }
        var out MalType
        var e error
        if out, e = rep(text); e != nil {
            if e.Error() == "<empty line>" { continue }
            fmt.Printf("Error: %v\n", e)
            continue
        }
        fmt.Printf("%v\n", out)
    }
}

File: impls/rust.2/step5_tco.rs
extern crate rustyline;

use rustyline::error::ReadlineError;
use rustyline::Editor;

mod core;
mod env;
mod printer;
mod reader;
mod types;

use std::rc::Rc;
use types::MalValue::MalList;
use types::{MalError, MalResult, MalValue};

fn read(input: String) -> Result<Option<MalValue>, MalError> {
    return reader::Reader::read_str(input.to_string());
}
fn eval(mut ast: MalValue, mut env: Rc<env::Environment>) -> MalResult {
    let result: MalResult;

    'tco: loop {
        result = match ast.clone() {
            MalValue::MalList(list) => {
                if list.len() == 0 {
                    return Ok(ast.clone());
                }
                let first_element = &list[0];
                match first_element {
                    MalValue::MalSymbol(ref s) if s == "def!" => {
                        return env.set(list[1].clone(), eval(list[2].clone(), env.clone())?);
                    }
                    MalValue::MalSymbol(ref s) if s == "let*" => {
                        assert_eq!(3, list.len());
                        let bindings = list[1].clone();
                        let expr = list[2].clone();

                        let new_env = Rc::new(env::Environment::new(Some(env.clone()), None, None));

                        if let Some(binding_list) = bindings.as_vec() {
                            assert!(binding_list.len() % 2 == 0);
                            let mut key = binding_list.first().unwrap();
                            for i in 0..binding_list.len() {
                                if i % 2 == 0 {
                                    key = binding_list.get(i).unwrap();
                                } else {
                                    new_env.set(
                                        key.clone(),
                                        eval(
                                            binding_list.get(i).unwrap().clone(),
                                            new_env.clone(),
                                        )?,
                                    )?;
                                }
                            }

                            ast = expr;
                            env = new_env;
                            continue 'tco;
                        } else {
                            return Err(MalError::EvalError(String::from("Missing bindings!")));
                        }
                    }
                    MalValue::MalSymbol(ref s) if s == "fn*" => {
                        return Ok(MalValue::MalFunc {
                            eval: eval,
                            ast: Rc::new(list[2].clone()),
                            env: env.clone(),
                            params: Rc::new(list[1].clone()),
                        });
                    }
                    MalValue::MalSymbol(ref s) if s == "do" => {
                        for token in &list[1..list.len() - 1] {
                            eval_ast(token.clone(), env.clone())?;
                        }

                        ast = list.last().unwrap().clone();
                        continue 'tco;
                    }
                    MalValue::MalSymbol(ref s) if s == "if" => {
                        assert!(list.len() >= 3);
                        let condition = eval(list[1].clone(), env.clone())?;
                        let true_expr = list[2].clone();
                        if condition.is_truthy() {
                            ast = true_expr.clone();
                            continue 'tco;
                        } else if list.len() == 4 {
                            ast = list[3].clone();
                            continue 'tco;
                        }
                        return Ok(MalValue::MalNil);
                    }
                    _ => match eval_ast(ast.clone(), env.clone())? {
                        MalList(list) => match list[0] {
                            MalValue::MalFunction(f, _) => return f(list[1..].to_vec()),
                            MalValue::MalFunc {
                                ast: ref fn_ast,
                                env: ref fn_env,
                                ref params,
                                ..
                            } => {
                                let a = &**fn_ast;
                                ast = a.clone();
                                env = Rc::new(env::Environment::new(
                                    Some(fn_env.clone()),
                                    params.as_vec(),
                                    Some(list[1..].to_vec()),
                                ));
                                continue 'tco;
                            }
                            _ => {
                                return Err(MalError::EvalError(
                                    "Attempt to call non-function".to_string(),
                                ))
                            }
                        },
                        _ => {
                            return Err(MalError::EvalError(String::from("Expected a list!")));
                        }
                    },
                }
            }
            _ => {
                return eval_ast(ast, env);
            }
        };

        break;
    }

    return result;
}

fn eval_ast(ast: MalValue, env: Rc<env::Environment>) -> Result<MalValue, MalError> {
    match ast.clone() {
        MalValue::MalSymbol(s) => {
            if let Ok(func) = env.get(ast.clone()) {
                return Ok(func);
            }
            return Err(types::MalError::EvalError(format!(
                "Symbol '{}' not found",
                s
            )));
        }
        MalValue::MalList(list) => {
            let mut result = Vec::<types::MalValue>::new();

            for token in list {
                result.push(eval(token, env.clone())?);
            }

            return Ok(types::MalValue::MalList(result));
        }
        MalValue::MalVector(vector) => {
            let mut result = Vec::<types::MalValue>::new();

            for token in vector {
                result.push(eval(token, env.clone())?);
            }

            return Ok(types::MalValue::MalVector(result));
        }
        MalValue::MalHashmap(keys, values) => {
            let mut result = Vec::<types::MalValue>::new();

            for token in values {
                result.push(eval(token, env.clone())?);
            }

            return Ok(types::MalValue::MalHashmap(keys, result));
        }
        _ => {
            return Ok(ast);
        }
    }
}

fn print(input: MalValue) -> String {
    return crate::printer::pr_str(input, true);
}

fn rep(input: String, env: &Rc<env::Environment>) -> String {
    let ast = read(input);
    match ast {
        Err(e) => {
            return format!("Error! {:?}", e);
        }
        Ok(v) => match v {
            None => return String::default(),
            Some(a) => match eval(a, env.clone()) {
                Ok(result) => {
                    return print(result);
                }
                Err(e) => {
                    return format!("Error! {:?}", e);
                }
            },
        },
    }
}

fn main() {
    let mut rl = Editor::<()>::new();
    if rl.load_history(".mal-history").is_err() {
        println!("No previous history.");
    }

    let env = Rc::new(env::Environment::new(None, None, None));

    for (symbol, func) in crate::core::ns() {
        if let Err(e) = env.set(MalValue::MalSymbol(symbol.to_string()), func) {
            panic!("Failed add core item {}. {:#?}", symbol, e)
        }
    }

    rep("(def! not (fn* (a) (if a false true)))".to_string(), &env);
    rep(
        "(def! sum2 (fn* (n acc) (if (= n 0) acc (sum2 (- n 1) (+ n acc)))))".to_string(),
        &env,
    );

    loop {
        let readline = rl.readline("user> ");
        match readline {
            Ok(line) => {
                rl.add_history_entry(line.as_str());
                let result = rep(line.trim_end().to_string(), &env);
                print!("{}", result);
                println!();
            }
            Err(ReadlineError::Interrupted) => break,
            Err(ReadlineError::Eof) => break,
            Err(err) => {
                println!("Error: {:?}", err);
                break;
            }
        }
    }
}
